receiver.go

```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
 		}
 	}
 
+	// for ModeFirst, auto-accept the message
+	modeFirstAccept := func(ctx context.Context, msg *Message) (*Message, error) {
+		if r.manualCreditor != nil || receiverSettleModeValue(r.link.ReceiverSettleMode) == ModeSecond {
+			return msg, nil
+		}
+		if err := r.AcceptMessage(ctx, msg); err != nil {
+			return nil, err
+		}
+		return msg, nil
+	}
+
 	// non-blocking receive to ensure buffered messages are
 	// delivered regardless of whether the link has been closed.
 	select {
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() non blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return modeFirstAccept(ctx, &msg)
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	default:
+		// done draining messages
 	}
 
 	// wait for the next message
 	select {
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return modeFirstAccept(ctx, &msg)
+	case <-r.link.close:
+		return nil, r.link.err
 	case <-r.link.Detached:
 		return nil, r.link.err
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	}
 }
 
+// Accept notifies the server that the message has been
+// accepted and does not require redelivery.
+func (r *Receiver) AcceptMessage(ctx context.Context, msg *Message) error {
+	if !msg.shouldSendDisposition() {
+		return nil
+	}
+	if err := r.messageDisposition(ctx, msg, &encoding.StateAccepted{}); err != nil {
+		return err
+	}
+	msg.settled = true
+	return nil
+}
+
+// Reject notifies the server that the message is invalid.
+//
+// Rejection error is optional.
+func (r *Receiver) RejectMessage(ctx context.Context, msg *Message, e *Error) error {
```

Comment:
Is there a reason it has to take an 'error'? Shouldn't it just be a string?

Reply:
+1 for using a string "reason" or something like this.

Reply:
This is how the method originally was and conforms to the [spec](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-rejected).  If we do this, then there is some info callers can't specify.

The .NET [impl](https://github.com/Azure/azure-amqp/blob/59894e8d992c63dfcaad6d1d3f454d9d88855d91/src/ReceivingAmqpLink.cs#L363) follows the same pattern, so I'm inclined to leave it.

All that said, what about adding a convenience func to create the appropriate `Error` from a reason string or the like?

Reply:
Leave it as is in this PR.